### ISSUE TITLE
telemetry: sql-stats subtest is skipped

### ISSUE THREAD

**Issue by yuzefovich** [cockroachdb/cockroach#107593]
Once #107555 merges, only `sql-stats` subtest of telemetry tests will be skipped. This is because the test currently fails with:
```
    datadriven.go:143: 
        testdata/telemetry/sql-stats:29:
         CREATE TABLE foo.t (x INT, y INT);
        SET application_name = '$ some other app';
        CREATE DATABASE bar;
        CREATE TABLE bar.t (x INT, y INT);
        SET application_name = 'hashed app name';
        CREATE DATABASE baz;
        CREATE TABLE baz.t (x INT, y INT);
        INSERT INTO baz.t VALUES (1, 1), (2, 2);
        output didn't match expected:
        @@ -1,13 +1,14 @@
         sql-stats
        - ├── $ some app
        - │    ├── [nodist] CREATE TABLE _ (_ INT8, _ INT8)
        - │    └── [nodist] SET application_name = '_'
        - ├── $ some other app
        - │    ├── [nodist] CREATE DATABASE _
        - │    ├── [nodist] CREATE TABLE _ (_ INT8, _ INT8)
        - │    └── [nodist] SET application_name = '_'
        + ├── 0a651942
        + │    └── [nodist] SELECT _ FROM _ WHERE _ = $1
        + ├── ad046cc2
        + │    └── [nodist] SELECT _ FROM _ WHERE _ = $1
          └── b16e4363
               ├── [nodist] CREATE DATABASE _
        +      ├── [nodist] CREATE DATABASE _
               ├── [nodist] CREATE TABLE _ (_ INT8, _ INT8)
        -      └── [nodist] INSERT INTO _ VALUES (_, _), (__more1_10__)
        +      ├── [nodist] CREATE TABLE _ (_ INT8, _ INT8)
        +      ├── [nodist] CREATE TABLE _ (_ INT8, _ INT8)
        +      ├── [nodist] INSERT INTO _ VALUES (_, _), (__more1_10__)
        +      └── [nodist] SET application_name = '_'
         
```

Jira issue: CRDB-30120

**Comment by ericharmeling** [comment]
It looks like there are couple things going on here:

1. The last application name set in a transaction is the one used when reporting statement stats for all statements in that transaction.

    For example, if we use explicit transactions:
    
    ```
    sql-stats
    BEGIN;
    CREATE TABLE foo.t (x INT, y INT);
    SET application_name = '$ some other app';
    COMMIT;
    BEGIN;
    CREATE DATABASE bar;
    CREATE TABLE bar.t (x INT, y INT);
    SET application_name = 'hashed app name';
    COMMIT;
    BEGIN;
    CREATE DATABASE baz;
    CREATE TABLE baz.t (x INT, y INT);
    INSERT INTO baz.t VALUES (1, 1), (2, 2);
    COMMIT;
    ----
     ├── $ some other app
     │    ├── [nodist] CREATE TABLE _ (_ INT8, _ INT8)
     │    └── [nodist] SET application_name = '_'
     ├── 0a651942
     │    └── [nodist] SELECT _ FROM _ WHERE _ = $1
     ├── ad046cc2
     │    └── [nodist] SELECT _ FROM _ WHERE _ = $1
     └── b16e4363
          ├── [nodist] CREATE DATABASE _
          ├── [nodist] CREATE DATABASE _
          ├── [nodist] CREATE TABLE _ (_ INT8, _ INT8)
          ├── [nodist] CREATE TABLE _ (_ INT8, _ INT8)
          ├── [nodist] INSERT INTO _ VALUES (_, _), (__more1_10__)
          └── [nodist] SET application_name = '_'
    ```
    
    Even the `SET application_name` statement is placed in the app name bucket *of the application name it is setting*. 

    Is this the expected behavior? I don't think so. If I run the same statements in `demo`, without explicit transactions, I get the expected app names in the `crdb_internal.cluster_statement_statistics` table:
    
    ```
    demo@127.0.0.1:26257/demoapp/movr> select crdb_internal.reset_sql_stats();
                                    ->
                                    -> SET application_name = '$ some app';
                                    -> CREATE DATABASE foo;
                                    -> CREATE TABLE foo.t (x INT, y INT);
                                    -> SET application_name = '$ some other app';
                                    -> CREATE DATABASE bar;
                                    -> CREATE TABLE bar.t (x INT, y INT);
                                    -> SET application_name = 'hashed app name';
                                    -> CREATE DATABASE baz;
                                    -> CREATE TABLE baz.t (x INT, y INT);
                                    -> INSERT INTO baz.t VALUES (1, 1), (2, 2);
                                    ->
                                    -> SELECT app_name, metadata ->> 'query' from crdb_internal.cluster_statement_statistics;
      crdb_internal.reset_sql_stats
    ---------------------------------
                    t
    (1 row)
    
    Time: 109ms total (execution 109ms / network 1ms)
    
    SET
    
    Time: 0ms total (execution 0ms / network 0ms)
    
    CREATE DATABASE
    
    Time: 1ms total (execution 1ms / network 0ms)
    
    CREATE TABLE
    
    Time: 2ms total (execution 2ms / network 0ms)
    
    SET
    
    Time: 0ms total (execution 0ms / network 0ms)
    
    CREATE DATABASE
    
    Time: 2ms total (execution 2ms / network 0ms)
    
    CREATE TABLE
    
    Time: 2ms total (execution 2ms / network 0ms)
    
    SET
    
    Time: 0ms total (execution 0ms / network 0ms)
    
    CREATE DATABASE
    
    Time: 1ms total (execution 1ms / network 0ms)
    
    CREATE TABLE
    
    Time: 2ms total (execution 2ms / network 0ms)
    
    INSERT 0 2
    
    Time: 3ms total (execution 2ms / network 0ms)
    
                  app_name             |       
    -----------------------------------+-----
    ...
      $ some app                       | CREATE DATABASE foo
      $ some app                       | CREATE TABLE foo.t (x INT8, y INT8)
      $ some app                       | SET application_name = '_'
      $ some app                       | SHOW database
      $ some other app                 | CREATE DATABASE bar
      $ some other app                 | CREATE TABLE bar.t (x INT8, y INT8)
      $ some other app                 | SET application_name = '_'
      $ some other app                 | SHOW database
      $$ $ cockroach demo              | SELECT username FROM system.users WHERE username = $1
      $$ $ some app                    | SELECT username FROM system.users WHERE username = $1
      $$ $ some other app              | SELECT username FROM system.users WHERE username = $1
      $$ hashed app name               | SELECT username FROM system.users WHERE username = $1
      $$ hashed app name               | SELECT username, privileges, grant_options FROM system.privileges WHERE path = $1
      hashed app name                  | CREATE DATABASE baz
      hashed app name                  | CREATE TABLE baz.t (x INT8, y INT8)
      hashed app name                  | INSERT INTO baz.t VALUES (_, _), (__more1_10__)
      hashed app name                  | SHOW database
    ...
    ```

    Shouldn't the in-memory statement stats look the same as the reported stats?

2. Somewhere, something is executing a `SELECT` statement when `CREATE DATABASE` is run. I believe this is an internal permissions/integrity check executed internally on behalf of a specific application (see the `SELECT username FROM system.users WHERE username = $1` statements above), and we aren't ignoring the sql stats generated by a designated application (prefixed `$$ `).

    We can *almost* "fix" this behavior by updating the [scrub check](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/conn_executor.go#L736-L739) to look for designated app names. The problem here is that if the designated app name also needs to be scrubbed, we still get the hashed app name, and we can't check if the app name is designated or not after it's hashed:
    
    ```
    sql-stats
     ├── $ some other app
     │    ├── [nodist] CREATE TABLE _ (_ INT8, _ INT8)
     │    └── [nodist] SET application_name = '_'
     ├── ad046cc2
     │    └── [nodist] SELECT _ FROM _ WHERE _ = $1
     └── b16e4363
          ├── [nodist] CREATE DATABASE _
          ├── [nodist] CREATE DATABASE _
          ├── [nodist] CREATE TABLE _ (_ INT8, _ INT8)
          ├── [nodist] CREATE TABLE _ (_ INT8, _ INT8)
          ├── [nodist] INSERT INTO _ VALUES (_, _), (__more1_10__)
          └── [nodist] SET application_name = '_'
    ```
    
    One thing we can do is hash *just the app name* when scrubbing statements executed by designated internal apps (i.e., 				`stat.Key.App = catconstants.DelegatedAppNamePrefix + HashForReporting(salt, stat.Key.App)`). This would allow us to check the designated prefix in the telemetry test (and elsewhere if we ever need to) after the statements have been scrubbed:
    
    ```
    sql-stats
     ├── $ some other app
     │    ├── [nodist] CREATE TABLE _ (_ INT8, _ INT8)
     │    └── [nodist] SET application_name = '_'
     └── b16e4363
          ├── [nodist] CREATE DATABASE _
          ├── [nodist] CREATE DATABASE _
          ├── [nodist] CREATE TABLE _ (_ INT8, _ INT8)
          ├── [nodist] CREATE TABLE _ (_ INT8, _ INT8)
          ├── [nodist] INSERT INTO _ VALUES (_, _), (__more1_10__)
          └── [nodist] SET application_name = '_'
    ```

    If that all looks fine, I'm happy to push the change to the scrub checks. But I don't think that will be sufficient to fix this test, as this test is meant to "`# Verify statements are correctly separated by app.`".

----

In summary: I'm a little stuck here. The stats collector/reporter in the test server isn't properly mapping statement stats to application names, and I'm not sure why.

**Comment by ericharmeling** [comment]
@yuzefovich I might need some help from SQL folks on this. Does this look like a problem with the way the test server reports statistics, or with the reporter itself? I can keep investigating myself if you think this is solidly obs-infra territory. :)

**Comment by yuzefovich** [comment]
Sorry, I don't have much context to add since I've never looked closely at the code in question. Perhaps someone form @cockroachdb/sql-foundations would be more helpful, maybe @rafiss?
